### DISCUSSION TITLE
Language Selection

### DISCUSSION THREAD

**Discussion by PascalBana** [Damcore-Studios/.github#1] [Ideas]
We need to come to a decision on what language will be used to build out most/all of the back end of the RE app. Due to the nature of the 
application, speed and safety are desired, JC has offered up four languages for us to test out, **Elixir**, **Go**, **Rust**, and **OCaml**. 

We will start by testing these languages and any others that are thought worth testing, then discuss any issues we think we might with each of them. One of our main goals is to ensure our codebase is easily maintainable. Given that this project will span years as well as limited current man power we would like to ensure that we can spend as much time as possible building new features rather than having to maintain and update old ones. Functional programming, it would seem, can help achieve this. 

Of course nothing said here should be read as stone. The purpose of opening this discussion is to prove any mislead arguments false and ensure we are following the best path for the applications development. for the time being we are foregoing the decision of what front end language will be used for the GUI. After the rest of the back end tech stack is decided we will start having discussion about the front end stack.

**Comment by PascalBana** [comment]
I'm going to start off by bashing on OCaml, after hearing about its promise of being a good programming language for financial apps, I went ahead and did some research. Most of all OCaml has issues integrating with certain services, I have read articles stating companies had wanted to change DB's or cloud service providers, only to be met with a hoard of issues as there were no stable setups for OCaml, large companies can take that because they have large development teams and most of the time OCaml is used in very static environments(Tech Stacks not changing very often). New organizations(Us) could find the lack of compatibility with certain services difficult(although not impossible) to deal with. on the bright i have seen that OCaml in comparison to rust allows some more fine tuning as well as simpler type control( I admittedly have doubts this will be necessary). Beyond this I don't have much to say about OCaml. Simply put if OCaml is to be used it could very well limit our options for the rest of the tech stack.

**Reply by JohnCarterGonzalez** [reply]
Yeah that was my main concern as well with Ocaml. I am not too opposed to writing the  bindings ourselves, (it would remove the headache for transient dependencies). We have the time and more libraries are being written everyday it seems. Also I would give a little push back on Type systems not being necessary, but that is a longer conversation and has been debated a lot. Ocaml needs some fleshing out as a general purpose language but Type system is a little too good, to discount simply for lack of libraries if we have the time. Food for thought.

**Comment by JohnCarterGonzalez** [comment]
Also we should keep an eye on the serde crate drama, as at the moment with it being a precompiled binary, there are obvious supply chain attacks, so much so that Fedora itself has halted use of serde until those concerns are addressed.  See here for more info: https://github.com/serde-rs/serde/issues/2538
